### ISSUE TITLE
NettyResponseFuture can wait indefinitely?

### ISSUE THREAD

**Issue by jon-ruckwood** [AsyncHttpClient/async-http-client#1070]
A http request was made days ago, the thread dump indicates that it was waiting - there may have been an unknown issue with the upstream server when the request was in progress:

```
Foo-1 id=19 state=WAITING
    - waiting on <0x55f017db> (a java.util.concurrent.CountDownLatch$Sync)
    - locked <0x55f017db> (a java.util.concurrent.CountDownLatch$Sync)
    at sun.misc.Unsafe.park(Native Method)
    at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
    at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
    at com.ning.http.client.providers.netty.future.NettyResponseFuture.get(NettyResponseFuture.java:152)
    ...
```
Nothing in the logs indicates an error/exception was encountered by async-http-client (logging is set to `INFO`). 

I am using version `1.9.15` and I am creating the client as such:

```java
AsyncHttpClient client = new AsyncHttpClient();
```

I appreciate there's not much to go on but do you know what may have happened here? 

I'm considering using `NettyResponseFuture#get(long l, TimeUnit tu)` instead but I would have thought that something would have aborted processing given the time involved? 

Thanks in advance.

**Comment by jon-ruckwood** [comment]
It seems like this can happen when the connection is closed by the server:

```java
@Test
public void asyncHttpClient() throws Exception {
    // given
    final MockWebServer server = new MockWebServer();

    server.enqueue(new MockResponse()
            .setBody("hello world")
            .setSocketPolicy(SocketPolicy.DISCONNECT_AT_START)
    );

    final String serverURL = server.url("/").toString();
    final AsyncHttpClient asyncHttpClient = new AsyncHttpClient();
    final Future<Response> future = asyncHttpClient.prepareGet(serverURL).execute();

    // when
    final Response response = future.get(); // waits indefinitely...?
}
```
Is this expected? 

Thanks.

**Comment by slandelle** [comment]
There was a bug in 2.0.0-RC5 that was cause AHC to choke on this case, I just fixed it.
I haven't tried with latest 1.9 yet.

Then, AHC retried by default (4 times by default, see config), and I wasn't able to get MockWebServer answer to the retry requests and I finally got a request time out (after 60s, which is the default).

**Comment by slandelle** [comment]
> Then, AHC retried by default (4 times by default, see config), and I wasn't able to get MockWebServer answer to the retry requests and I finally got a request time out (after 60s, which is the default).

My bad, the default is 5, and I only enqueued 4 responses. It get the expected crash.

**Comment by slandelle** [comment]
Marking as fixed for AHC2.
Contributions welcome for getting this fixed on AHC1, I personally don't have the cycles to investigate on the old code base, sorry.

**Comment by jon-ruckwood** [comment]
I've switched to using the latest AHC2 RC. Thanks for helping me get to the bottom of this!

**Comment by slandelle** [comment]
Beware that the fix hasn't been released yet :)
It will be in RC6 (end of this week or the next one).

Thanks for reporting, and sorry I really can't afford trying to fix AHC 1.9.
